### PR TITLE
Add string_utils::to_string()

### PR DESCRIPTION
**Pull request by vbkaisetsu**
ANDROID's libstdc++ does not have `std::to_string()`.

This branch adds `string_utils::to_string()` functions that simulates behavior of `std::to_string()` on ANDROID.

### CODE REVIEW

**Review by odashi — CHANGES_REQUESTED** [review]
See following page and fix code.

https://cpprefjp.github.io/reference/string/to_string.html

Then we become not to need `std::to_string()` wherever we want to convert numericals to strings.

**Review by odashi — CHANGES_REQUESTED** [review]
And there is no test code.

**Comment by vbkaisetsu** [comment]
Fixed.

File: primitiv/string_utils.h [hunk 1]
```diff
@@ -31,34 +31,39 @@ inline std::string join(
 }
 
 /**
- * Wrapper functions of std::to_string()
+ * Imprementation of std::to_string()
+ *
+ * Some libstdc++ (e.g. Android) do not support std::to_string().
+ * These functions support them.
```
Comment (odashi) [inline primitiv/string_utils.h]: What is *them*?

File: primitiv/string_utils.h [hunk 2]
```diff
@@ -31,34 +31,39 @@ inline std::string join(
...
 inline std::string to_string(std::uint32_t value) {
-  char *c_str = nullptr;
-  if (asprintf(&c_str, "%d", value) < 0) {
-    THROW_ERROR("Failed to format value");
-  }
-  std::unique_ptr<char, decltype(&std::free)> u(c_str, std::free);
-  std::string s(u.get());
-  return s;
+  /*
+   * max uint32_t = 4294967295
```
Comment (odashi) [inline primitiv/string_utils.h]: Probably redundant. Line 45 has enough information.

File: primitiv/string_utils.h [hunk 3]
```diff
@@ -31,34 +31,39 @@ inline std::string join(
...
+   *                 ~~~~~~~~~
+   *                 digits10 = 9
+   *
+   * buffer's size = (digits10 + 1) + NULL
```
Comment (odashi) [inline primitiv/string_utils.h]: `NULL` can be used to only pointers. Should describe `'\0'` in this case.

File: primitiv/string_utils.h [hunk 4]
```diff
@@ -31,34 +31,39 @@ inline std::string join(
...
 inline std::string to_string(float value) {
-  char *c_str = nullptr;
-  if (asprintf(&c_str, "%f", value) < 0) {
-    THROW_ERROR("Failed to format value");
-  }
-  std::unique_ptr<char, decltype(&std::free)> u(c_str, std::free);
-  std::string s(u.get());
-  return s;
+  /*
+   * max float = 340282346638528859811704183484516925440.000000
```
Comment (odashi) [inline primitiv/string_utils.h]: ditto

File: test/string_utils_test.cc [hunk 5]
```diff
@@ -28,5 +28,24 @@ TEST_F(StringUtilsTest, CheckJoin) {
...
 }
 
+TEST_F(StringUtilsTest, CheckToString) {
+  // Generated by std::to_string()
+  EXPECT_EQ("0.000000", to_string(0.0f));
+  EXPECT_EQ("1.000000", to_string(1.0f));
+  EXPECT_EQ("1.100000", to_string(1.1f));
+  EXPECT_EQ("1.234568", to_string(1.23456789f));
+  EXPECT_EQ("340282346638528859811704183484516925440.000000",
+            to_string(340282346638528859811704183484516925440.0f)); // max float
```
Comment (odashi) [inline test/string_utils_test.cc]: Use `numeric_limits`.

**Review by odashi — CHANGES_REQUESTED** [review]

**Comment by vbkaisetsu** [comment]
Fixed.

File: primitiv/string_utils.h [hunk 6]
```diff
@@ -2,9 +2,11 @@
 #define PRIMITIV_STRING_UTILS_H_
 
 #include <cstdint>
+#include <cstdlib>
 #include <sstream>
 #include <string>
 #include <vector>
+#include <primitiv/error.h>
```
Comment (odashi) [inline primitiv/string_utils.h]: Unnecessary include.

**Comment by odashi** [comment]
Maybe it is better to implement `to_string()` of all types to avoid accidental implicit type conversions:

- `int` and `unsigned`
- `long` and `unsigned long`
- `long long` and `unsigned long long`
- `float`, `double` and `long double`

File: test/string_utils_test.cc [hunk 7]
```diff
@@ -28,5 +29,24 @@ TEST_F(StringUtilsTest, CheckJoin) {
...
+            to_string(std::numeric_limits<float>::max())); // max float
+  EXPECT_EQ("-1.000000", to_string(-1.0f));
+  EXPECT_EQ("-1.100000", to_string(-1.1f));
+  EXPECT_EQ("-12.345679", to_string(-12.3456789f));
+  EXPECT_EQ("-340282346638528859811704183484516925440.000000",
+            to_string(-std::numeric_limits<float>::max())); // min float
+  EXPECT_EQ("0", to_string(0u));
+  EXPECT_EQ("5", to_string(5u));
+  EXPECT_EQ("123", to_string(123u));
+  EXPECT_EQ("4294967295", to_string(0xffffffffu)); // max uint32_t
```
Comment (odashi) [inline test/string_utils_test.cc]: Also should be `numeric_limits`.

**Comment by vbkaisetsu** [comment]
I wrote behaviour of `to_string()` for all types including `long double`. However, the size of `long double` may be different on architectures, so I wrote test cases of `to_string(long double)` only for some 64bit systems.

**Comment by odashi** [comment]
Usually, **all** floating point types (including even `float` and `double`) are *implementation-defined*, and we can not write always correct tests for such types with hard-coded constants. I think the most reliable way of tests is determining length of resulting strings using `max_exponent10`.

File: primitiv/string_utils.h [hunk 8]
```diff
@@ -28,6 +30,51 @@ std::string join(
...
 }
 
+/**
+ * Imprementation of std::to_string()
+ *
+ * Some libstdc++ (e.g. Android) do not support std::to_string().
+ * We support libraries that do not have std::to_string().
+ */
+
+#define DEF_TO_STRING(type, format, size) \
```
Comment (odashi) [inline primitiv/string_utils.h]: I don't recommended to use macros in general headers.

**Review by odashi — CHANGES_REQUESTED** [review]

File: test/string_utils_test.cc [hunk 9]
```diff
@@ -28,5 +29,240 @@ TEST_F(StringUtilsTest, CheckJoin) {
...
+  EXPECT_EQ("2147483647",
+            to_string(std::numeric_limits<std::int32_t>::max()));
+  EXPECT_EQ("-2147483648",
+            to_string(std::numeric_limits<std::int32_t>::min()));
+  EXPECT_EQ("0", to_string(0u));
+  EXPECT_EQ("5", to_string(5u));
+  EXPECT_EQ("123", to_string(123u));
+  EXPECT_EQ("4294967295",
+            to_string(std::numeric_limits<std::uint32_t>::max()));
+#if __x86_64__ || __ppc64__
```
Comment (odashi) [inline test/string_utils_test.cc]: Use `#if defined`.

File: test/string_utils_test.cc [hunk 10]
```diff
@@ -28,5 +29,240 @@ TEST_F(StringUtilsTest, CheckJoin) {
...
+  EXPECT_EQ("9223372036854775807",
+            to_string(std::numeric_limits<std::int64_t>::max()));
+  EXPECT_EQ("-9223372036854775808",
+            to_string(std::numeric_limits<std::int64_t>::min()));
+  EXPECT_EQ("0", to_string(0ull));
+  EXPECT_EQ("5", to_string(5ull));
+  EXPECT_EQ("123", to_string(123ull));
+  EXPECT_EQ("18446744073709551615",
+            to_string(std::numeric_limits<std::uint64_t>::max()));
+#endif  // __x86_64__ || __ppc64__
```
Comment (odashi) [inline test/string_utils_test.cc]: Fix comment.

File: test/string_utils_test.cc [hunk 11]
```diff
@@ -28,5 +29,240 @@ TEST_F(StringUtilsTest, CheckJoin) {
...
+  EXPECT_EQ("-12.345679", to_string(-12.3456789));
+  EXPECT_EQ("-17976931348623157081452742373170435679807056752584499659891747"
+             "68031572607800285387605895586327668781715404589535143824642343"
+             "21326889464182768467546703537516986049910576551282076245490090"
+             "38932894407586850845513394230458323690322294816580855933212334"
+             "8274797826204144723168738177180919299881250404026184124858368."
+             "000000",
+            to_string(-std::numeric_limits<double>::max()));
+#if __x86_64__ || __ppc64__
+  /*
```
Comment (odashi) [inline test/string_utils_test.cc]: Redundant comment. The content is completely same as above condition.

File: test/string_utils_test.cc [hunk 12]
```diff
@@ -28,5 +29,240 @@ TEST_F(StringUtilsTest, CheckJoin) {
...
+  EXPECT_EQ("-1.100000", to_string(-1.1));
+  EXPECT_EQ("-12.345679", to_string(-12.3456789));
+  EXPECT_EQ("-17976931348623157081452742373170435679807056752584499659891747"
+             "68031572607800285387605895586327668781715404589535143824642343"
+             "21326889464182768467546703537516986049910576551282076245490090"
+             "38932894407586850845513394230458323690322294816580855933212334"
+             "8274797826204144723168738177180919299881250404026184124858368."
+             "000000",
+            to_string(-std::numeric_limits<double>::max()));
+#if __x86_64__ || __ppc64__
```
Comment (odashi) [inline test/string_utils_test.cc]: ditto

File: test/string_utils_test.cc [hunk 13]
```diff
@@ -28,5 +29,240 @@ TEST_F(StringUtilsTest, CheckJoin) {
...
+            "254511086154895839508779675546413794489596052797520987481383976"
+            "257859210575628440175934932416214833956535018919681138909184379"
+            "573470326940634289008780584694035245347939808067427323629788710"
+            "086717580253156130235606487870925986528841635097252953709111431"
+            "720488774740553905400942537542411931794417513706468964386151771"
+            "884986701034153254238591108962471088538580868883777725864856414"
+            "593426212108664758848926003176234596076950884914966244415660441"
+            "9552086811989770240.000000",
+            to_string(-std::numeric_limits<long double>::max()));
+#endif  // __x86_64__ || __ppc64__
```
Comment (odashi) [inline test/string_utils_test.cc]: ditto

**Review by odashi — CHANGES_REQUESTED** [review]

**Comment by odashi** [comment]
However current tests look enough for usual systems, and I'll merge this branch after solving comments.

File: test/string_utils_test.cc [hunk 14]
```diff
@@ -28,5 +29,240 @@ TEST_F(StringUtilsTest, CheckJoin) {
...
+#if __x86_64__ || __ppc64__
+  /*
+   * We guarantee behavior of to_string(long double) only for x86_64 and ppc64
+   * operating systems.
+   */
+  EXPECT_EQ("0.000000", to_string(0.0L));
+  EXPECT_EQ("1.000000", to_string(1.0L));
+  EXPECT_EQ("1.100000", to_string(1.1L));
+  EXPECT_EQ("1.234568", to_string(1.23456789L));
+  EXPECT_EQ("118973149535723176502126385303097020516906332229462420044032373"
```
Comment (odashi) [inline test/string_utils_test.cc]: This string literal is identical with that about the minimum value except the top `"-"` character, and both can be integrated into one string literal. Also `float` and `double` string literals can be integrated into one.

**Review by odashi — CHANGES_REQUESTED** [review]

File: primitiv/string_utils.h [hunk 15]
```diff
@@ -28,6 +30,51 @@ std::string join(
...
+  inline std::string to_string(type value) { \
+    char buffer[std::numeric_limits<type>::size]; \
```
Comment (odashi) [inline primitiv/string_utils.h]: This array consumes a huge space of the stack (especially when converting from `long double`) that is usually not recommended. But using dynamic arrays (e.g., `vector`) requires some heap allocations. I have no good idea to fix this.

**Review by odashi — APPROVED** [review]

**Pull request merged**